src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenForEachTests.cs

```diff
@@ -2365,5 +2365,696 @@ .locals init (T V_0)
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: @"1
+2
+3");
+        }
+
+        [Fact]
+        public void TestGetEnumeratorPatternViaExtensionsCSharp8()
+        {
+            string source = @"
+using System;
+public class C
+{
+    public static void Main()
+    {
+        foreach (var i in new C())
+        {
+            Console.WriteLine(i);
+        }
+    }
+    public sealed class Enumerator
+    {
+        public int Current { get; private set; }
+        public bool MoveNext() => Current++ != 3;
+    }
+}
+public static class Extensions
+{
+    public static C.Enumerator GetEnumerator(this C self) => new C.Enumerator();
+}";
+            var comp = CreateCompilationWithMscorlib46(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (7,27): error CS8652: The feature 'extension GetEnumerator' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         foreach (var i in new C())
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "new C()").WithArguments("extension GetEnumerator").WithLocation(7, 27)
+                );
+        }
+
+        [Fact]
+        public void TestGetEnumeratorPatternViaInternalExtensions()
+        {
+            string source = @"
+using System;
+public class C
+{
+    public static void Main()
+    {
+        foreach (var i in new C())
+        {
+            Console.WriteLine(i);
+        }
+    }
+    public sealed class Enumerator
+    {
+        public int Current { get; private set; }
+        public bool MoveNext() => Current++ != 3;
+    }
+}
+public static class Extensions
+{
+    internal static C.Enumerator GetEnumerator(this C self) => new C.Enumerator();
+}";
+            var comp = CreateCompilationWithMscorlib46(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (7,27): warning CS0279: 'C' does not implement the 'collection' pattern. 'Extensions.GetEnumerator(C)' is either static or not public.
+                //         foreach (var i in new C())
+                Diagnostic(ErrorCode.WRN_PatternStaticOrInaccessible, "new C()").WithArguments("C", "collection", "Extensions.GetEnumerator(C)").WithLocation(7, 27),
+                // (7,27): error CS1579: foreach statement cannot operate on variables of type 'C' because 'C' does not contain a public instance definition for 'GetEnumerator'
+                //         foreach (var i in new C())
+                Diagnostic(ErrorCode.ERR_ForEachMissingMember, "new C()").WithArguments("C", "GetEnumerator").WithLocation(7, 27)
+                );
+        }
+
+        [Fact]
+        public void TestGetEnumeratorPatternViaExtensionOnInternalClass()
+        {
+            string source = @"
+using System;
+public class C
+{
+    public static void Main()
+    {
+        foreach (var i in new C())
```

Comment:
Could you add a test where this is a `ref var` using `StructRefEnumerator<T> GetEnumerator<T>(this T[] array)`

Reply:
Another test idea: extension on tuples involving nested conversions (https://github.com/dotnet/roslyn/issues/16159)